lib/puppet/util/at_fork.rb

```diff
@@ -0,0 +1,35 @@
+require 'puppet'
+
+# A module for building AtFork handlers. These handlers are objects providing
+# pre/post fork callbacks modeled after those registered by the `pthread_atfork`
+# function.
+# Currently there are two AtFork handler implementations:
+# - a noop implementation used on all platforms except Solaris (and possibly
+#   even there as a fallback)
+# - a Solaris implementation which ensures the forked process runs in a different
+#   contract than the parent process. This is necessary for agent runs stared by
+#   the puppet agent service to be able to restart that service without being
+#   killed in the process as a consequence of running in the same contract as the
+#   service.
+module Puppet::Util::AtFork
+  @handler_class = loop do
+    if Facter.value(:operatingsystem) == 'Solaris'
```

Comment:
Thor vendored into bundler does that: https://github.com/bundler/bundler/blob/09358d459565c39314eea933020c8f848b3cfce0/lib/bundler/vendor/thor/lib/thor/shell/basic.rb#L378

One issue is it doesn't work in JRuby (but that doesn't apply here). Or maybe [`RbConfig::CONFIG['host_os']`](http://stackoverflow.com/questions/11784109/detecting-operating-systems-in-ruby)?